Maple2.Server.Game/PacketHandlers/ShopHandler.cs

```diff
@@ -21,20 +21,57 @@ private enum Command : byte {
         Sell = 5,
         InstantRestock = 9,
         Refresh = 10,
-        LoadNew = 13,
+        LoadNew = 13, // Possibly unable to function in game
     }
 
     public override void Handle(GameSession session, IByteReader packet) {
         var command = packet.Read<Command>();
         switch (command) {
+            case Command.PurchaseBuyBack:
+                HandlePurchaseBuyBack(session, packet);
+                return;
             case Command.Buy:
                 HandleBuy(session, packet);
                 return;
+            case Command.Sell:
+                HandleSell(session, packet);
+                return;
+            case Command.InstantRestock:
+                HandleInstantRestock(session, packet);
+                return;
+            case Command.Refresh:
+                HandleRefresh(session);
+                return;
         }
     }
 
+    private void HandlePurchaseBuyBack(GameSession session, IByteReader packet) {
+        int id = packet.ReadInt();
+
+        session.Shop.PurchaseBuyBack(id);
+    }
+
     private void HandleBuy(GameSession session, IByteReader packet) {
         int shopItemId = packet.ReadInt();
         int quantity = packet.ReadInt();
+
+        session.Shop.Buy(shopItemId, quantity);
+    }
+
+    private void HandleSell(GameSession session, IByteReader packet) {
+        long itemUid = packet.ReadLong();
+        int quantity = packet.ReadInt();
+
+        session.Shop.Sell(itemUid, quantity);
+    }
+
+    private void HandleInstantRestock(GameSession session, IByteReader packet) {
+        packet.ReadInt(); // cost
```

Comment:
missing code to charge the user?

Reply:
Nah. It's not using the packet read here to ensure the current price is correct using ShopManager.InstantRestock